### ISSUE TITLE
Using 2:1 Muxes instead of NAND gates as basic element

### ISSUE THREAD

**Issue by Michel-FK** [con-dog/hack_emulator_c#1]
You say:
- Start with a single software defined NAND gate in C
    - EVERYTHING is built from this base chip - everything
 
But then, I don't see any NAND gate in the muxes implementation?

Then, why not start with an even simpler element, like a 2:1 mux?
![Mux Logic Gates](https://github.com/user-attachments/assets/05584787-3db6-49a4-869e-0b0d6e4322ec)

**Comment by con-dog** [comment]
Hey @Michel-FK  I'm tracking close to the assignments from NandToTetris course, from NAND you can build any possible chip, and represent any possible boolean combination (just using NAND). Is this true with a 2:1 Mux? 

You'll find my primitve NAND gate inside components/combinatorial/logic-gates/bit/nand/nand.c

By "EVERYTHING is built from this base chip - everything" I mean I use the NAND gate as the starting point, and build other chips from it, then use those newly made chips to build other chips etc.

Technically you could swap everything out for NAND gates, but that would take a very long time to code up

**Comment by Michel-FK** [comment]
> from NAND you can build any possible chip, and represent any possible boolean combination (just using NAND). Is this true with a 2:1 Mux?

Yes, as shown in my previous picture, you can emulate all basic logic gates, and of course higher-degree muxes like 4:1.

But by turning a 4:1 mux upside-down (using selects as input and inputs as function value), you have a _programmable_ logic gate like on the last line in the picture: I use the 4:1 mux as a 2-input LUT (Look Up Table), so you can emulate any function with 2 inputs and 1 output.

And the C code to emulate a 2:1 mux is trivial:

```
static inline bool mux_2_1(bool a, bool b, bool s)
{
    return s ? b : a;
}
```

For emulating a NAND gate:
```
static inline bool nand(bool a, bool b)
{
    return mux_2_1(1, 0, mux_2_1(0, a, b));
}
```

For the more complex programmable logic unit:
```
static inline bool lut(a, b, f0, f1, f2, f3)
{
    return(mux_2_1(mux_2_1(f0, f1, a), mux_2_1(f2, f3, a), b);
}
```
Using the LUT as a programmable NAND gate:
```
y = lut(a, b, 1, 1, 1, 0);
```

**Comment by con-dog** [comment]
ahh thats cool, I understand now. Looks like a good approach. 

At least for my first pass of this emulator, I want to track close to the NandToTetris assignments, so someone coming from the course can see the course project implemented in C

**Comment by oneman** [comment]
Good Work.

Continue hacking ;)

/DR/
